clang/lib/Sema/SemaType.cpp

```diff
@@ -7963,6 +7967,148 @@ static Attr *getCCTypeAttr(ASTContext &Ctx, ParsedAttr &Attr) {
   llvm_unreachable("unexpected attribute kind!");
 }
 
+ExprResult Sema::ActOnEffectExpression(Expr *CondExpr, FunctionEffectMode &Mode,
+                                       bool RequireConstexpr) {
+  // see checkFunctionConditionAttr, Sema::CheckCXXBooleanCondition
+  if (RequireConstexpr || !CondExpr->isTypeDependent()) {
+    ExprResult E = PerformContextuallyConvertToBool(CondExpr);
+    if (E.isInvalid())
+      return E;
+    CondExpr = E.get();
+    if (RequireConstexpr || !CondExpr->isValueDependent()) {
+      llvm::APSInt CondInt;
+      E = VerifyIntegerConstantExpression(
```

Comment:
Should we be calling `ActOnFinishFullExpr` before we try and evaluate this? Alternatively, would it be possible to treat this as a converted constant expression?

CC @cor3ntin, @erichkeane, @AaronBallman

Reply:
Also, can you add some test cases with placeholder expressions as the argument, e.g. `[[clang::nonblocking(__builtin_memset)]]`, and also a case where the argument is an unexpanded pack?

Reply:
> Also, can you add some test cases with placeholder expressions as the argument, e.g. `[[clang::nonblocking(__builtin_memset)]]`, and also a case where the argument is an unexpanded pack?

Not quite sure how to construct a test that won't fail in uninteresting other ways ... e.g.

```
void t5() [[clang::nonblocking(__builtin_memset)]] {}
// error: nonblocking attribute requires an integer constant
// (via err_attribute_argument_type)

template <bool V>
struct ValueDependent {};

template <typename X>
struct TypeDependent {};

ValueDependent<__builtin_memset> x3;
// error: builtin functions must be directly called

TypeDependent<__builtin_memset> x3;
// error: template argument for template type parameter must be a type
```

I wasn't able to find examples of tests using unexpanded packs, at least not in Sema. Would appreciate any tips, thanks. I've got a few more basic tests done and will push those shortly.

Reply:
Thanks. With the unexpanded pack: During initial parsing of the function type, it's seen as dependent and represented that way, holding onto the condition.

Is there a way at this time to detect that the expression is not only dependent but an unexpanded pack? Maybe it's as simple as `E->containsUnexpandedParameterPack()`? That looks promising, but it's bugging me that there are no (obvious) precedents for this.

If I add `void g() { f<true, false>(); }` there is a second attempt to resolve the expression but it is still dependent, so clearly there is an error here. But I can imagine cases where it takes more than one level of template instantiation for a dependent expression to be resolved.

Reply:
> Is there a way at this time to detect that the expression is not only dependent but an unexpanded pack?

Iirc pack dependence is a thing. You should be able to check for that.

> If I add void g() { f<true, false>(); } there is a second attempt to resolve the expression but it is still dependent, so clearly there is an error here

I’ll have to check how we currently handle these cases because I’m not terribly familiar w/ packs myself.

Reply:
`Sema::ActOnEffectExpression()` is now much simpler -- it checks for dependence and unexpanded parameter packs before trying to evaluate the expression as an integer constant.